### PR TITLE
feat: fallback to build protobuf from source

### PR DESCRIPTION
**Pull request by tisonkun**

### CODE REVIEW

File: src/protobuf_impl.rs [hunk 1]
```diff
@@ -23,21 +23,8 @@ fn get_protoc() -> String {
...
         return s;
     }
-
     // The bundled protoc should always match the version
-    let protoc_bin_name = match (env::consts::OS, env::consts::ARCH) {
-        ("linux", "x86") => "protoc-linux-x86_32",
-        ("linux", "x86_64") => "protoc-linux-x86_64",
-        ("linux", "aarch64") => "protoc-linux-aarch_64",
-        ("linux", "powerpc64") => "protoc-linux-ppcle_64",
-        ("macos", "x86_64") => "protoc-osx-x86_64",
```
Comment (tisonkun) [inline src/protobuf_impl.rs:33]: The initial motivation is to support macos + aarch64. We can bundle the new binary but why?

**Comment by BusyJay** [comment]
I'm afraid building from source would require extra dependencies like C/C++ compilers.

**Comment by tisonkun** [comment]
> I'm afraid building from source would require extra dependencies like C/C++ compilers.

This patch include three changes:

1. Support new protoc version string.
2. Prost build respects bundled protoc
3. Using protobuf-src over binaries to support macos aarch64 fallback

If you'd prefer add macos aarch64 binary to resolve 3 and agree on the changes for 1 and 2, I can find time to restore and update the binaries.

File: src/lib.rs [hunk 2]
```diff
@@ -15,11 +15,48 @@ mod protobuf_impl;
...
+
+fn check_protoc_version(protoc: &str) -> Result<String, ()> {
+    let ver_re = Regex::new(r"([0-9]+)\.([0-9]+)(\.[0-9])?").unwrap();
+    let output = Command::new(protoc).arg("--version").output();
+    match output {
+        Ok(o) => {
+            let caps = ver_re.captures(from_utf8(&o.stdout).unwrap()).unwrap();
+            let major = caps.get(1).unwrap().as_str().parse::<i16>().unwrap();
+            let minor = caps.get(2).unwrap().as_str().parse::<i16>().unwrap();
+            if (major == 3 && minor >= 1) || (major >= 20) {
```
Comment (BusyJay) [inline src/lib.rs]: How about `(major, minor) >= (3, 1)`?

**Review by BusyJay — COMMENTED** [review]
I believe protoc-osx-aarch64 was omitted because it exceeded the size limit (10MiB at that time) of a crate that can be published to crates.io. Better check if the problem still exists.

**Comment by BusyJay** [comment]
Note protobuf-src may not work for Windows.

File: src/lib.rs (hunk 2 above)
Comment (tisonkun) [reply src/lib.rs]: See the link in commit https://github.com/tikv/protobuf-build/pull/67/commits/c1462c6644a23123105b4508ddeecd64c8457ba9.

I tried to follow protobuf's version policy that they change the version pattern only after 20.x. But it should be physically OK to use `(major, minor) >= (3, 1)`.

**Comment by tisonkun** [comment]
> Note protobuf-src may not work for Windows.

OK. Then I'll try the restore and update binaries way. Perhaps I can run `cargo publish --dry-run` locally to check the limitation but I'm not sure if dry-run requires push permission to the crate.

**Comment by tisonkun** [comment]
@BusyJay Updated. I use 23.3 protoc and check it works well on osx-aarch_64 env.

**Review by BusyJay — APPROVED** [review]

**Comment by BusyJay** [comment]
The resulting package is clearly larger than 10MiB, I'm afraid this change can't be published to crates.io.

**Comment by tisonkun** [comment]
```
cargo publish --dry-run
    Updating crates.io index
   Packaging protobuf-build v0.14.1 (/Users/tison/Brittani/protobuf-build)
   Verifying protobuf-build v0.14.1 (/Users/tison/Brittani/protobuf-build)
    Updating crates.io index
   Compiling memchr v2.5.0
   Compiling protobuf v2.28.0
   Compiling bytes v1.4.0
   Compiling regex-syntax v0.7.2
   Compiling bitflags v1.3.2
   Compiling aho-corasick v1.0.2
   Compiling regex v1.8.4
   Compiling protobuf-codegen v2.28.0
   Compiling protobuf-build v0.14.1 (/Users/tison/Brittani/protobuf-build/target/package/protobuf-build-0.14.1)
    Finished dev [unoptimized + debuginfo] target(s) in 13.22s
    Packaged 31 files, 53.8MiB (17.7MiB compressed)
   Uploading protobuf-build v0.14.1 (/Users/tison/Brittani/protobuf-build)
warning: aborting upload due to dry run
```

It seems dry-run doesn't test against it.

cc @nrc could you please take a look and check the cargo limitation?

cc @weihanglo

**Comment by weihanglo** [comment]
crates.io has its own stricter limitations that Cargo doesn't check at this time. Part of the reason Cargo has less checks because Cargo is not only for crates.io. We would like to see a new publish check API but the work is not on priority so far I can tell.

The 10MiB limit is documented [here](https://doc.rust-lang.org/nightly/cargo/reference/publishing.html#packaging-a-crate).

See also

* https://github.com/rust-lang/cargo/issues/4300
* https://github.com/rust-lang/crates-io-cargo-teams/issues/82

**Comment by tisonkun** [comment]
@BusyJay Perhaps we drop the support to `linux-ppcle_64` and `linux-x86_32` (or perhaps `linux-aarch_64`, I don't know) as they are least widely used? And if we can add a feature flag to support users to fallback to protobuf-src on their demand.

```
-rwxr-xr-x@  1 tison  staff   7.8M Jan  1  1980 protoc-linux-aarch_64
-rwxr-xr-x@  1 tison  staff   9.1M Jan  1  1980 protoc-linux-ppcle_64
-rwxr-xr-x@  1 tison  staff   8.5M Jan  1  1980 protoc-linux-x86_32
-rwxr-xr-x@  1 tison  staff   7.8M Jan  1  1980 protoc-linux-x86_64
-rwxr-xr-x@  1 tison  staff   6.1M Jun 14 23:12 protoc-osx-aarch_64
-rwxr-xr-x@  1 tison  staff   5.6M Jun 14 23:12 protoc-osx-x86_64
-rwxr-xr-x@  1 tison  staff   8.8M Jan  1  1980 protoc-win32.exe

-rw-r--r--  1 tison  staff    18M Jun 21 12:56 target/package/protobuf-build-0.14.1.crate
```

**Comment by tisonkun** [comment]
@weihanglo ... and Thank you!

**Comment by BusyJay** [comment]
I think fallback to build from source may not be a bad idea giving that the community prefers not to use bundled binaries (like prost-build recently did). But we still need to think a way to support Windows.

**Comment by BusyJay** [comment]
linux-ppcle_64 and linux-aarch_64 are mainly used on server side.

**Comment by tisonkun** [comment]
@BusyJay Updated. As we discussed offline -

1. Bundle win only;
2. Others compile from source.

**Comment by tisonkun** [comment]
> linux-ppcle_64 and linux-aarch_64 are mainly used on server side.

Yeah. I suspect on server side they're used less, while aarch_64 can be a trend :D

**Review by hawkingrei — APPROVED** [review]

File: src/lib.rs [hunk 3]
```diff
@@ -53,7 +103,7 @@ impl Builder {
     }
 
     pub fn include_google_protos(&mut self) -> &mut Self {
-        let path = format!("{}/include", std::env!("CARGO_MANIFEST_DIR"));
+        let path = format!("{}/include", env!("CARGO_MANIFEST_DIR"));
```
Comment (BusyJay) [inline src/lib.rs:106]: Maybe protos in include directory also need to be updated.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

File: src/lib.rs (hunk 3 above)
Comment (tisonkun) [reply src/lib.rs:106]: Updated.

**Review by BusyJay — APPROVED** [review]

**Review by sunxiaoguang — APPROVED** [review]
LGTM

**Comment by tisonkun** [comment]
Maybe we can release a 0.15.0 now and upgrade downsteams like raft-rs and tikv.

**Pull request merged**